src/garage/torch/algos/pearl.py

```diff
@@ -1,6 +1,635 @@
+            self._qf2.parameters(),
+            lr=qf_lr,
+        )
+        self.vf_optimizer = optimizer_class(
+            self._vf.parameters(),
+            lr=vf_lr,
+        )
+        self.context_optimizer = optimizer_class(
+            self.policy.networks[0].parameters(),
+            lr=context_lr,
+        )
+
+    def __getstate__(self):
+        """Object.__getstate__.
+
+        Returns:
+            dict: the state to be pickled for the instance.
+
+        """
+        data = self.__dict__.copy()
+        del data['_replay_buffers']
+        del data['_context_replay_buffers']
+        del data['evaluator']
+        return data
+
+    def train(self, runner):
+        """Obtain samples, train, and evaluate for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        """
+        for _ in runner.step_epochs():
+            epoch = runner.step_itr / self._num_steps_per_epoch
+
+            # obtain initial set of samples from all train tasks
+            if epoch == 0:
+                for idx in range(self._num_train_tasks):
+                    self._task_idx = idx
+                    self.obtain_samples(runner, epoch, self._num_initial_steps,
+                                        np.inf)
+
+            # obtain samples from random tasks
+            for _ in range(self._num_tasks_sample):
+                idx = np.random.randint(self._num_train_tasks)
+                self._task_idx = idx
+                self._context_replay_buffers[idx].clear()
+                # obtain samples with z ~ prior
+                if self._num_steps_prior > 0:
+                    self.obtain_samples(runner, epoch, self._num_steps_prior,
+                                        np.inf)
+                # obtain samples with z ~ posterior
+                if self._num_steps_posterior > 0:
+                    self.obtain_samples(runner, epoch,
+                                        self._num_steps_posterior,
+                                        self._update_post_train)
+                # obtain extras samples for RL training but not encoder
+                if self._num_extra_rl_steps_posterior > 0:
+                    self.obtain_samples(runner,
+                                        epoch,
+                                        self._num_extra_rl_steps_posterior,
+                                        self._update_post_train,
+                                        add_to_enc_buffer=False)
+
+            logger.log('Training...')
+            # sample train tasks and optimize networks
+            self.train_once()
+            runner.step_itr += 1
+
+            logger.log('Evaluating...')
+            # evaluate
+            self.policy.reset_belief()
+            self.evaluator.evaluate(self)
+
+    def train_once(self):
```

Comment:
can you make this (and all other methods not part of `MetaRLAlgorithm`) private?

the new classmethods are an exception, of course.